File: libs/core/src/semantic/conversion_classifier.cpp
/**
 * @file conversion_classifier.cpp
 * @author Francesco Corso
 * @date 8/9/2026
 * @brief
 */
module;

#include <libassert/assert-macros.hpp>

module prism.core:semantic.conversion_classifier.impl;

import :semantic.conversion_classifier;
import :binder;
import :semantic.compilation;
import :symbols.named_type_symbol;

namespace prism
{
    Conversion ConversionClassifier::classify_conversion(const TypeSymbol &source, const TypeSymbol &destination) const
    {
        if (&source == &destination)
        {
            return get_trivial_conversion(ConversionKind::identity);
        }

        if (source.special_type() == destination.special_type() && source.special_type() != SpecialType::none)
            [[unlikely]]
        {
            return get_trivial_conversion(ConversionKind::identity);
        }

        if (is_numeric_type(source) && is_numeric_type(destination))
        {
            return classify_numeric_conversion(source.special_type(), destination.special_type());
        }

        if (is_character_type(source) && is_character_type(destination))
        {
            return classify_character_conversion(source.special_type(), destination.special_type());
        }

        return no_conversion;
    }

    Optional<OperandConversion> ConversionClassifier::classify_unary_operand_type(const UnaryOperation operation,
                                                                                  const TypeSymbol &operand) const
    {
        switch (operation)
        {
            case UnaryOperation::identity:
                if (is_numeric_type(operand))
                {
                    auto &promoted = promote_numeric_type(operand);
                    return OperandConversion{
                        .conversion = classify_numeric_conversion(operand.special_type(), promoted.special_type()),
                        .type = promoted};
                }
                break;
            case UnaryOperation::negation:
                if (is_numeric_type(operand))
                {
                    return promote_negation_type(operand).transform(
                        [&](const TypeSymbol &promoted)
                        {
                            return OperandConversion{.conversion = classify_numeric_conversion(operand.special_type(),
                                                                                               promoted.special_type()),
                                                     .type = promoted};
                        });
                }
                break;
            case UnaryOperation::logical_not:
                if (operand.special_type() == SpecialType::bool_)
                {
                    return OperandConversion{.conversion = get_trivial_conversion(ConversionKind::identity),
                                             .type = operand};
                }
                break;
            case UnaryOperation::bitwise_not:
                if (is_integral_type(operand))
                {
                    auto &promoted = promote_numeric_type(operand);
                    return OperandConversion{
                        .conversion = classify_numeric_conversion(operand.special_type(), promoted.special_type()),
                        .type = promoted};
                }
                break;
            case UnaryOperation::pre_increment:
            case UnaryOperation::pre_decrement:
            case UnaryOperation::post_increment:
            case UnaryOperation::post_decrement:
                if (is_numeric_type(operand))
                {
                    return OperandConversion{.conversion = get_trivial_conversion(ConversionKind::identity),
                                             .type = operand};
                }
                break;
        }

        return std::nullopt;
    }

    Optional<BinaryOperandConversion> ConversionClassifier::classify_binary_operand_type(
        const BinaryOperation operation,
        const TypeSymbol &left,
        const TypeSymbol &right) const
    {
        const auto get_numeric_conversion = [&](const TypeSymbol &promoted)
        {
            return BinaryOperandConversion{
                .left_conversion = classify_numeric_conversion(left.special_type(), promoted.special_type()),
                .right_conversion = classify_numeric_conversion(right.special_type(), promoted.special_type()),
                .type = promoted};
        };

        switch (operation)
        {
            case BinaryOperation::addition:
            case BinaryOperation::subtraction:
            case BinaryOperation::multiplication:
            case BinaryOperation::division:
            case BinaryOperation::modulo:
                if (is_numeric_type(left) && is_numeric_type(right))
                {
                    return get_common_numeric_type(left, right).transform(get_numeric_conversion);
                }
                break;
            case BinaryOperation::bitwise_and:
            case BinaryOperation::bitwise_or:
            case BinaryOperation::bitwise_xor:
            case BinaryOperation::shift_left:
            case BinaryOperation::shift_right:
            case BinaryOperation::unsigned_shift_right:
                if (is_integral_type(left) && is_integral_type(right))
                {
                    return get_common_numeric_type(left, right).transform(get_numeric_conversion);
                }
                break;
            case BinaryOperation::logical_and:
            case BinaryOperation::logical_or:
                if (left.special_type() == SpecialType::bool_ && right.special_type() == SpecialType::bool_)
                {
                    return BinaryOperandConversion{.left_conversion = get_trivial_conversion(ConversionKind::identity),
                                                   .right_conversion = get_trivial_conversion(ConversionKind::identity),
                                                   .type = left};
                }
                break;
            case BinaryOperation::equals:
            case BinaryOperation::not_equals:
                if (left.special_type() == SpecialType::bool_ && right.special_type() == SpecialType::bool_)
                {
                    return BinaryOperandConversion{.left_conversion = get_trivial_conversion(ConversionKind::identity),
                                                   .right_conversion = get_trivial_conversion(ConversionKind::identity),
                                                   .type = left};
                }

                if (is_numeric_type(left) && is_numeric_type(right))
                {
                    return get_common_numeric_type(left, right).transform(get_numeric_conversion);
                }

                if (is_character_type(left) && is_numeric_type(right))
                {
                    auto &promoted = get_common_character_type(left, right);
                    return BinaryOperandConversion{
                        .left_conversion = classify_character_conversion(left.special_type(), promoted.special_type()),
                        .right_conversion = classify_numeric_conversion(promoted.special_type(), right.special_type()),
                        .type = promoted};
                }

                break;
            case BinaryOperation::less_than:
            case BinaryOperation::less_than_or_equals:
            case BinaryOperation::greater_than:
            case BinaryOperation::greater_than_or_equals:
            case BinaryOperation::three_way_comparison:
                if (is_numeric_type(left) && is_numeric_type(right))
                {
                    return get_common_numeric_type(left, right).transform(get_numeric_conversion);
                }
                break;
            case BinaryOperation::null_coalescing:
                throw NotImplementedException{};
        }

        UNREACHABLE("Invalid binary operation");
    }

    const Compilation &ConversionClassifier::compilation() const noexcept
    {
        return binder_.compilation();
    }

    bool ConversionClassifier::is_numeric_type(const TypeSymbol &type) noexcept
    {
        return is_numeric(type.special_type());
    }

    bool ConversionClassifier::is_integral_type(const TypeSymbol &type) noexcept
    {
        return is_integer(type.special_type());
    }

    Conversion ConversionClassifier::classify_numeric_conversion(const SpecialType source,
                                                                 const SpecialType destination) const noexcept
    {
        if (source == destination)
        {
            return get_trivial_conversion(ConversionKind::identity);
        }

        if (is_implicit_numeric_conversion(source, destination))
        {
            return get_trivial_conversion(ConversionKind::implicit_numeric);
        }

        return get_trivial_conversion(ConversionKind::explicit_numeric);
    }

    bool ConversionClassifier::is_implicit_numeric_conversion(const SpecialType source,
                                                              const SpecialType destination) const noexcept
    {
        const NumericInfo source_info = numeric_info(source);
        const NumericInfo destination_info = numeric_info(destination);

        if (source_info.family == NumericFamily::none || destination_info.family == NumericFamily::none)
        {
            return false;
        }

        if ((source_info.family == NumericFamily::signed_integer &&
             destination_info.family == NumericFamily::signed_integer) ||
            (source_info.family == NumericFamily::unsigned_integer &&
             destination_info.family == NumericFamily::unsigned_integer))
        {
            return source_info.width <= destination_info.width;
        }

        if (source_info.family == NumericFamily::unsigned_integer &&
            destination_info.family == NumericFamily::signed_integer)
        {
            return source_info.width <= destination_info.width;
        }

        if (is_integer_family(source_info.family) && destination_info.family == NumericFamily::floating_point)
        {
            return source_info.precision <= destination_info.precision;
        }

        if (source_info.family == NumericFamily::floating_point &&
            destination_info.family == NumericFamily::floating_point)
        {
            return source_info.width <= destination_info.width;
        }

        return false;
    }

    ConversionClassifier::NumericInfo ConversionClassifier::numeric_info(const SpecialType type) const
    {
        switch (type)
        {
            case SpecialType::i8:
                return {.family = NumericFamily::signed_integer, .width = 8, .precision = 8};

            case SpecialType::i16:
                return {.family = NumericFamily::signed_integer, .width = 16, .precision = 16};

            case SpecialType::i32:
                return {.family = NumericFamily::signed_integer, .width = 32, .precision = 32};

            case SpecialType::i64:
                return {.family = NumericFamily::signed_integer, .width = 64, .precision = 64};

            case SpecialType::i128:
                return {.family = NumericFamily::signed_integer, .width = 128, .precision = 128};

            case SpecialType::isize:
                {
                    auto [pointer_width] = compilation().target_settings();
                    return {.family = NumericFamily::signed_integer,
                            .width = pointer_width,
                            .precision = pointer_width};
                }

            case SpecialType::u8:
                return {.family = NumericFamily::unsigned_integer, .width = 8, .precision = 8};

            case SpecialType::u16:
                return {.family = NumericFamily::unsigned_integer, .width = 16, .precision = 16};

            case SpecialType::u32:
                return {.family = NumericFamily::unsigned_integer, .width = 32, .precision = 32};

            case SpecialType::u64:
                return {.family = NumericFamily::unsigned_integer, .width = 64, .precision = 64};

            case SpecialType::u128:
                return {.family = NumericFamily::unsigned_integer, .width = 128, .precision = 128};

            case SpecialType::usize:
                {
                    auto [pointer_width] = compilation().target_settings();
                    return {.family = NumericFamily::unsigned_integer,
                            .width = pointer_width,
                            .precision = pointer_width};
                }

            case SpecialType::f32:
                return {.family = NumericFamily::floating_point, .width = 32, .precision = 24};

            case SpecialType::f64:
                return {.family = NumericFamily::floating_point, .width = 64, .precision = 53};

            default:
                return {};
        }
    }

    bool ConversionClassifier::is_integer_family(const NumericFamily family) noexcept
    {
        return family == NumericFamily::signed_integer || family == NumericFamily::unsigned_integer;
    }

    const TypeSymbol &ConversionClassifier::promote_numeric_type(const TypeSymbol &source) const
    {
        DEBUG_ASSERT(is_numeric(source.special_type()));
        switch (source.special_type())
        {
            case SpecialType::i8:
            case SpecialType::i16:
            case SpecialType::u8:
            case SpecialType::u16:
                return compilation().get_special_type(SpecialType::i32);
            default:
                return source;
        }
    }

    Optional<const TypeSymbol &> ConversionClassifier::promote_negation_type(const TypeSymbol &source) const
    {
        DEBUG_ASSERT(is_numeric(source.special_type()));

        switch (source.special_type())
        {
            case SpecialType::i8:
            case SpecialType::i16:
            case SpecialType::u8:
            case SpecialType::u16:
                return compilation().get_special_type(SpecialType::i32);

            case SpecialType::u32:
                return compilation().get_special_type(SpecialType::i64);

            case SpecialType::u64:
                return compilation().get_special_type(SpecialType::i128);

            case SpecialType::usize:
                switch (compilation().target_settings().pointer_width)
                {
                    case 32:
                        return compilation().get_special_type(SpecialType::i64);
                    case 64:
                        return compilation().get_special_type(SpecialType::i128);
                    default:
                        throw InvalidStateException{"Invalid pointer width"};
                }

            case SpecialType::u128:
                return std::nullopt;

            default:
                return source;
        }
    }
    Optional<const TypeSymbol &> ConversionClassifier::get_common_numeric_type(const TypeSymbol &left,
                                                                               const TypeSymbol &right) const
    {
        DEBUG_ASSERT(is_numeric(left.special_type()) && is_numeric(right.special_type()));

        if (left.special_type() == right.special_type())
            return promote_numeric_type(left);

        auto left_family = numeric_info(left.special_type()).family;
        auto right_family = numeric_info(right.special_type()).family;

        if (left_family == right_family)
        {
            return promote_numeric_type(wider_precision(left, right));
        }

        if (is_integer_family(left_family) && right_family == NumericFamily::floating_point)
        {
            return common_float_type(right, left);
        }

        if (left_family == NumericFamily::floating_point && is_integer_family(right_family))
        {
            return common_float_type(left, right);
        }

        if (left_family == NumericFamily::signed_integer && right_family == NumericFamily::unsigned_integer)
        {
            return common_signed_type(left, right)
                .transform([this](const TypeSymbol &symbol) -> auto & { return promote_numeric_type(symbol); });
        }

        if (left_family == NumericFamily::unsigned_integer && right_family == NumericFamily::signed_integer)
        {
            return common_signed_type(right, left)
                .transform([this](const TypeSymbol &symbol) -> auto & { return promote_numeric_type(symbol); });
        }

        return std::nullopt;
    }

    const TypeSymbol &ConversionClassifier::wider_precision(const TypeSymbol &left, const TypeSymbol &right) const
    {
        const auto left_info = numeric_info(left.special_type());
        const auto right_info = numeric_info(right.special_type());
        DEBUG_ASSERT(left_info.family == right_info.family);

        if (left_info.precision < right_info.precision)
            return right;

        return left;
    }

    Optional<const TypeSymbol &> ConversionClassifier::common_float_type(const TypeSymbol &float_type,
                                                                         const TypeSymbol &integer_type) const
    {
        const auto float_info = numeric_info(float_type.special_type());
        const auto integer_info = numeric_info(integer_type.special_type());
        DEBUG_ASSERT(float_info.family == NumericFamily::floating_point);
        DEBUG_ASSERT(is_integer_family(integer_info.family));

        if (float_info.precision <= integer_info.precision)
            return float_type;

        if (float_type.special_type() == SpecialType::f32 &&
            integer_info.precision <= numeric_info(SpecialType::f64).precision)
        {
            return compilation().get_special_type(SpecialType::f64);
        }

        return std::nullopt;
    }

    Optional<const TypeSymbol &> ConversionClassifier::common_signed_type(const TypeSymbol &signed_type,
                                                                          const TypeSymbol &unsigned_type) const
    {
        const auto signed_info = numeric_info(signed_type.special_type());
        const auto unsigned_info = numeric_info(unsigned_type.special_type());
        DEBUG_ASSERT(signed_info.family == NumericFamily::signed_integer);
        DEBUG_ASSERT(unsigned_info.family == NumericFamily::unsigned_integer);

        if (signed_info.precision > unsigned_info.precision)
            return signed_type;

        switch (unsigned_type.special_type())
        {
            case SpecialType::u8:
                return compilation().get_special_type(SpecialType::i16);
            case SpecialType::u16:
                return compilation().get_special_type(SpecialType::i32);
            case SpecialType::u32:
                return compilation().get_special_type(SpecialType::i64);
            case SpecialType::u64:
                return compilation().get_special_type(SpecialType::i128);
            case SpecialType::u128:
                return std::nullopt;
            case SpecialType::usize:
                switch (compilation().target_settings().pointer_width)
                {
                    case 32:
                        return compilation().get_special_type(SpecialType::i64);
                    case 64:
                        return compilation().get_special_type(SpecialType::i128);
                    default:
                        throw InvalidStateException{"Invalid pointer width"};
                }
            default:
                UNREACHABLE("This isn't a valid path");
        }
    }

    bool ConversionClassifier::is_character_type(const TypeSymbol &type) noexcept
    {
        return is_character(type.special_type());
    }

    std::int32_t ConversionClassifier::character_width(const SpecialType type) noexcept
    {
        DEBUG_ASSERT(is_character(type));
        switch (type)
        {
            case SpecialType::char_:
                return 8;
            case SpecialType::char16:
                return 16;
            case SpecialType::rune:
                return 32;
            default:
                std::unreachable();
        }
    }

    Conversion ConversionClassifier::classify_character_conversion(const SpecialType source,
                                                                   const SpecialType destination) noexcept
    {
        if (source == destination)
            return get_trivial_conversion(ConversionKind::identity);

        const auto source_width = character_width(source);
        const auto destination_width = character_width(destination);
        DEBUG_ASSERT(source_width != destination_width);

        if (source_width > destination_width)
            return get_trivial_conversion(ConversionKind::explicit_character);

        return get_trivial_conversion(ConversionKind::implicit_character);
    }

    const TypeSymbol &ConversionClassifier::get_common_character_type(const TypeSymbol &left,
                                                                      const TypeSymbol &right) noexcept
    {
        if (&left == &right)
            return left;

        const auto source_width = character_width(left.special_type());
        const auto destination_width = character_width(right.special_type());
        DEBUG_ASSERT(source_width != destination_width);

        if (source_width > destination_width)
            return left;

        return right;
    }

} // namespace prism

File: src/Prism.Core/Semantic/ConversionClassifier.cs
using System.Diagnostics;
using Prism.Core.Binding;
using Prism.Core.Compiling;
using Prism.Core.Configuration;
using Prism.Core.Symbols;

namespace Prism.Core.Semantic;

internal readonly record struct OperandConversion(Conversion Conversion, TypeSymbol Type);

internal readonly record struct BinaryOperandConversion(
    Conversion LeftConversion,
    Conversion RightConversion,
    TypeSymbol Type
);

internal sealed class ConversionClassifier(Binder binder)
{
    public Conversion ClassifyConversion(TypeSymbol source, TypeSymbol target)
    {
        if (source == target)
        {
            return Conversion.GetTrivial(ConversionKind.Identity);
        }

        if (IsNumericType(source) && IsNumericType(target))
        {
            return ClassifyNumericConversion(source.SpecialType, target.SpecialType);
        }

        if (IsCharacterType(source) && IsCharacterType(target))
        {
            return ClassifyCharacterConversion(source.SpecialType, target.SpecialType);
        }

        return Conversion.None;
    }

    public OperandConversion? ClassifyUnaryOperand(UnaryOperation operation, TypeSymbol operand)
    {
        switch (operation)
        {
            case UnaryOperation.Identity:
                if (IsNumericType(operand))
                {
                    var promoted = PromoteNumericType(operand);
                    return new OperandConversion(
                        ClassifyNumericConversion(operand.SpecialType, promoted.SpecialType),
                        promoted
                    );
                }
                break;
            case UnaryOperation.Negation:
                if (IsNumericType(operand) && PromoteNegationType(operand) is { } negated)
                {
                    return new OperandConversion(
                        ClassifyNumericConversion(operand.SpecialType, negated.SpecialType),
                        negated
                    );
                }
                break;
            case UnaryOperation.LogicalNot:
                if (operand.SpecialType == SpecialType.Bool)
                {
                    return new OperandConversion(
                        Conversion.GetTrivial(ConversionKind.Identity),
                        operand
                    );
                }
                break;
            case UnaryOperation.BitwiseNot:
                if (IsIntegralType(operand))
                {
                    var promoted = PromoteNumericType(operand);
                    return new OperandConversion(
                        ClassifyNumericConversion(operand.SpecialType, promoted.SpecialType),
                        promoted
                    );
                }
                break;
            case UnaryOperation.PreIncrement:
            case UnaryOperation.PreDecrement:
            case UnaryOperation.PostIncrement:
            case UnaryOperation.PostDecrement:
                if (IsNumericType(operand))
                {
                    return new OperandConversion(
                        Conversion.GetTrivial(ConversionKind.Identity),
                        operand
                    );
                }
                break;
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(operation), operation, null);
        }

        return null;
    }

    public BinaryOperandConversion? ClassifyBinaryOperand(
        BinaryOperation operation,
        TypeSymbol left,
        TypeSymbol right
    )
    {
        switch (operation)
        {
            case BinaryOperation.Add:
            case BinaryOperation.Subtract:
            case BinaryOperation.Multiply:
            case BinaryOperation.Divide:
            case BinaryOperation.Modulo:
                if (IsNumericType(left) && IsNumericType(right))
                {
                    return GetNumericConversion(GetCommonNumericType(left, right));
                }
                break;
            case BinaryOperation.BitwiseAnd:
            case BinaryOperation.BitwiseOr:
            case BinaryOperation.BitwiseXor:
            case BinaryOperation.ShiftLeft:
            case BinaryOperation.ShiftRight:
            case BinaryOperation.UnsignedShiftRight:
                if (IsIntegralType(left) && IsIntegralType(right))
                {
                    return GetNumericConversion(GetCommonNumericType(left, right));
                }
                break;
            case BinaryOperation.LogicalAnd:
            case BinaryOperation.LogicalOr:
                if (left.SpecialType == SpecialType.Bool && right.SpecialType == SpecialType.Bool)
                {
                    return new BinaryOperandConversion(
                        Conversion.GetTrivial(ConversionKind.Identity),
                        Conversion.GetTrivial(ConversionKind.Identity),
                        binder.Compilation.GetSpecialType(SpecialType.Bool)
                    );
                }
                break;
            case BinaryOperation.Equal:
            case BinaryOperation.NotEqual:
                if (left.SpecialType == SpecialType.Bool && right.SpecialType == SpecialType.Bool)
                {
                    return new BinaryOperandConversion(
                        Conversion.GetTrivial(ConversionKind.Identity),
                        Conversion.GetTrivial(ConversionKind.Identity),
                        left
                    );
                }

                if (IsNumericType(left) && IsNumericType(right))
                {
                    return GetNumericConversion(GetCommonNumericType(left, right));
                }

                if (IsCharacterType(left) && IsCharacterType(right))
                {
                    var promoted = GetCommonCharacterType(left, right);
                    return new BinaryOperandConversion(
                        Conversion.GetTrivial(ConversionKind.Identity),
                        Conversion.GetTrivial(ConversionKind.Identity),
                        promoted
                    );
                }

                break;
            case BinaryOperation.LessThan:
            case BinaryOperation.LessThanOrEqual:
            case BinaryOperation.GreaterThan:
            case BinaryOperation.GreaterThanOrEqual:
            case BinaryOperation.ThreeWayComparison:
                if (IsNumericType(left) && IsNumericType(right))
                {
                    return GetNumericConversion(GetCommonNumericType(left, right));
                }
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(operation), operation, null);
        }

        return null;

        BinaryOperandConversion? GetNumericConversion(TypeSymbol? promoted)
        {
            if (promoted is null)
                return null;

            return new BinaryOperandConversion(
                ClassifyNumericConversion(left.SpecialType, promoted.SpecialType),
                ClassifyNumericConversion(right.SpecialType, promoted.SpecialType),
                promoted
            );
        }
    }

    private Compilation Compilation => binder.Compilation;

    private static bool IsNumericType(TypeSymbol type)
    {
        return type.SpecialType.IsNumeric;
    }

    private static bool IsIntegralType(TypeSymbol type)
    {
        return type.SpecialType.IsInteger;
    }

    private Conversion ClassifyNumericConversion(SpecialType source, SpecialType destination)
    {
        if (source == destination)
        {
            return Conversion.GetTrivial(ConversionKind.Identity);
        }

        return Conversion.GetTrivial(
            IsImplicitNumericConversion(source, destination)
                ? ConversionKind.ImplicitNumeric
                : ConversionKind.ExplicitNumeric
        );
    }

    private bool IsImplicitNumericConversion(SpecialType source, SpecialType destination)
    {
        var sourceInfo = GetNumericInfo(source);
        var destinationInfo = GetNumericInfo(destination);

        if (sourceInfo.Family == NumericFamily.None || destinationInfo.Family == NumericFamily.None)
        {
            return false;
        }

        if (
            (
                sourceInfo.Family == NumericFamily.SignedInteger
                && destinationInfo.Family == NumericFamily.SignedInteger
            )
            || (
                sourceInfo.Family == NumericFamily.UnsignedInteger
                && destinationInfo.Family == NumericFamily.UnsignedInteger
            )
        )
        {
            return sourceInfo.Width <= destinationInfo.Width;
        }

        if (
            sourceInfo.Family == NumericFamily.UnsignedInteger
            && destinationInfo.Family == NumericFamily.SignedInteger
        )
        {
            return sourceInfo.Width < destinationInfo.Width;
        }

        if (
            IsIntegerFamily(sourceInfo.Family)
            && destinationInfo.Family == NumericFamily.FloatingPoint
        )
        {
            return sourceInfo.Precision <= destinationInfo.Precision;
        }

        if (
            sourceInfo.Family == NumericFamily.FloatingPoint
            && destinationInfo.Family == NumericFamily.FloatingPoint
        )
        {
            return sourceInfo.Width <= destinationInfo.Width;
        }

        return false;
    }

    private NumericInfo GetNumericInfo(SpecialType type)
    {
        return type switch
        {
            SpecialType.I8 => new NumericInfo(NumericFamily.SignedInteger, 8),
            SpecialType.I16 => new NumericInfo(NumericFamily.SignedInteger, 16),
            SpecialType.I32 => new NumericInfo(NumericFamily.SignedInteger, 32),
            SpecialType.I64 => new NumericInfo(NumericFamily.SignedInteger, 64),
            SpecialType.I128 => new NumericInfo(NumericFamily.SignedInteger, 128),
            SpecialType.ISize => new NumericInfo(
                NumericFamily.SignedInteger,
                Compilation.Settings.PointerWidth.BitWidth
            ),
            SpecialType.U8 => new NumericInfo(NumericFamily.UnsignedInteger, 8),
            SpecialType.U16 => new NumericInfo(NumericFamily.UnsignedInteger, 16),
            SpecialType.U32 => new NumericInfo(NumericFamily.UnsignedInteger, 32),
            SpecialType.U64 => new NumericInfo(NumericFamily.UnsignedInteger, 64),
            SpecialType.U128 => new NumericInfo(NumericFamily.UnsignedInteger, 128),
            SpecialType.USize => new NumericInfo(
                NumericFamily.UnsignedInteger,
                Compilation.Settings.PointerWidth.BitWidth
            ),
            SpecialType.F32 => new NumericInfo(NumericFamily.FloatingPoint, 32, 24),
            SpecialType.F64 => new NumericInfo(NumericFamily.FloatingPoint, 64, 53),
            _ => default,
        };
    }

    private static bool IsIntegerFamily(NumericFamily family)
    {
        return family is NumericFamily.SignedInteger or NumericFamily.UnsignedInteger;
    }

    private TypeSymbol PromoteNumericType(TypeSymbol source)
    {
        Debug.Assert(source.SpecialType.IsNumeric);
        return source.SpecialType switch
        {
            SpecialType.I8 or SpecialType.I16 or SpecialType.U8 or SpecialType.U16 =>
                Compilation.GetSpecialType(SpecialType.I32),
            _ => source,
        };
    }

    private TypeSymbol? PromoteNegationType(TypeSymbol source)
    {
        Debug.Assert(source.SpecialType.IsNumeric);
        return source.SpecialType switch
        {
            SpecialType.I8 or SpecialType.I16 or SpecialType.U8 or SpecialType.U16 =>
                Compilation.GetSpecialType(SpecialType.I32),
            SpecialType.U32 => Compilation.GetSpecialType(SpecialType.I64),
            SpecialType.U64 => Compilation.GetSpecialType(SpecialType.I128),
            SpecialType.USize => Compilation.Settings.PointerWidth switch
            {
                PointerWidth.X32 => Compilation.GetSpecialType(SpecialType.I64),
                PointerWidth.X64 => Compilation.GetSpecialType(SpecialType.I128),
                _ => throw new InvalidOperationException("Invalid pointer width"),
            },
            SpecialType.U128 => null,
            _ => source,
        };
    }

    private TypeSymbol? GetCommonNumericType(TypeSymbol left, TypeSymbol right)
    {
        Debug.Assert(left.SpecialType.IsNumeric && right.SpecialType.IsNumeric);

        if (left.SpecialType == right.SpecialType)
            return PromoteNumericType(left);

        var leftFamily = GetNumericInfo(left.SpecialType).Family;
        var rightFamily = GetNumericInfo(right.SpecialType).Family;

        if (leftFamily == rightFamily)
            return PromoteNumericType(WiderPrecision(left, right));

        if (IsIntegerFamily(leftFamily) && rightFamily == NumericFamily.FloatingPoint)
        {
            return CommonFloatType(right, left);
        }

        if (leftFamily == NumericFamily.FloatingPoint && IsIntegerFamily(rightFamily))
        {
            return CommonFloatType(left, right);
        }

        if (
            leftFamily == NumericFamily.SignedInteger
            && rightFamily == NumericFamily.UnsignedInteger
        )
        {
            var commonSigned = CommonSignedType(left, right);
            return commonSigned is not null ? PromoteNumericType(commonSigned) : null;
        }

        // ReSharper disable once InvertIf
        if (
            leftFamily == NumericFamily.UnsignedInteger
            && rightFamily == NumericFamily.SignedInteger
        )
        {
            var commonSigned = CommonSignedType(right, left);
            return commonSigned is not null ? PromoteNumericType(commonSigned) : null;
        }

        return null;
    }

    private TypeSymbol WiderPrecision(TypeSymbol left, TypeSymbol right)
    {
        var leftInfo = GetNumericInfo(left.SpecialType);
        var rightInfo = GetNumericInfo(right.SpecialType);
        Debug.Assert(leftInfo.Family == rightInfo.Family);

        return leftInfo.Precision < rightInfo.Precision ? right : left;
    }

    private TypeSymbol? CommonFloatType(TypeSymbol floatType, TypeSymbol integerType)
    {
        var floatInfo = GetNumericInfo(floatType.SpecialType);
        var integerInfo = GetNumericInfo(integerType.SpecialType);
        Debug.Assert(floatInfo.Family == NumericFamily.FloatingPoint);
        Debug.Assert(IsIntegerFamily(integerInfo.Family));

        if (floatInfo.Precision >= integerInfo.Precision)
            return floatType;

        if (
            floatType.SpecialType == SpecialType.F32
            && integerInfo.Precision <= GetNumericInfo(SpecialType.F64).Precision
        )
        {
            return Compilation.GetSpecialType(SpecialType.F64);
        }

        return null;
    }

    private TypeSymbol? CommonSignedType(TypeSymbol signedType, TypeSymbol unsignedType)
    {
        var signedInfo = GetNumericInfo(signedType.SpecialType);
        var unsignedInfo = GetNumericInfo(unsignedType.SpecialType);
        Debug.Assert(signedInfo.Family == NumericFamily.SignedInteger);
        Debug.Assert(unsignedInfo.Family == NumericFamily.UnsignedInteger);

        if (signedInfo.Precision > unsignedInfo.Precision)
            return signedType;

        return unsignedType.SpecialType switch
        {
            SpecialType.U8 => Compilation.GetSpecialType(SpecialType.I16),
            SpecialType.U16 => Compilation.GetSpecialType(SpecialType.I32),
            SpecialType.U32 => Compilation.GetSpecialType(SpecialType.I64),
            SpecialType.U64 => Compilation.GetSpecialType(SpecialType.I128),
            SpecialType.U128 => (TypeSymbol?)null,
            SpecialType.USize => Compilation.Settings.PointerWidth switch
            {
                PointerWidth.X32 => Compilation.GetSpecialType(SpecialType.I64),
                PointerWidth.X64 => Compilation.GetSpecialType(SpecialType.I128),
                _ => throw new InvalidOperationException("Invalid pointer width"),
            },
            _ => throw new InvalidOperationException("Invalid unsigned type"),
        };
    }

    private static bool IsCharacterType(TypeSymbol type)
    {
        return type.SpecialType.IsCharacter;
    }

    private static int CharacterWidth(SpecialType type)
    {
        Debug.Assert(type.IsCharacter);
        return type switch
        {
            SpecialType.Char => 8,
            SpecialType.Char16 => 16,
            SpecialType.Rune => 32,
            _ => throw new InvalidOperationException("Invalid character type"),
        };
    }

    private static Conversion ClassifyCharacterConversion(
        SpecialType source,
        SpecialType destination
    )
    {
        if (source == destination)
            return Conversion.GetTrivial(ConversionKind.Identity);

        var sourceWidth = CharacterWidth(source);
        var destinationWidth = CharacterWidth(destination);
        Debug.Assert(sourceWidth != destinationWidth);

        return sourceWidth > destinationWidth
            ? Conversion.GetTrivial(ConversionKind.ExplicitCharacter)
            : Conversion.GetTrivial(ConversionKind.ImplicitCharacter);
    }

    private static TypeSymbol GetCommonCharacterType(TypeSymbol left, TypeSymbol right)
    {
        if (left == right)
            return left;

        var sourceWidth = CharacterWidth(left.SpecialType);
        var destinationWidth = CharacterWidth(right.SpecialType);
        Debug.Assert(sourceWidth != destinationWidth);

        return sourceWidth > destinationWidth ? left : right;
    }

    private enum NumericFamily : byte
    {
        None,
        SignedInteger,
        UnsignedInteger,
        FloatingPoint,
    };

    private readonly record struct NumericInfo(NumericFamily Family, int Width, int Precision)
    {
        public NumericInfo(NumericFamily family, int width)
            : this(family, width, width) { }
    }
}
